lib/service.dart

```diff
@@ -192,6 +232,243 @@ class IsolateManager {
+        },
+      );
+    }
+  }
+
+  void resetAvailableExtensions() {
+    firstFrameEventReceived = false;
+    _serviceExtensions.clear();
+    _serviceExtensionController
+        .forEach((String name, StreamController<bool> stream) {
+      stream.add(false);
+    });
+  }
+
+  /// Sets the state for a service extension and makes the call to the VMService.
+  void setServiceExtensionState(String name, bool enabled, dynamic value) {
+    _callServiceExtension(name, value);
+
+    final StreamController<ServiceExtensionState> streamController =
+        _getServiceExtensionStateController(name);
+    streamController.add(new ServiceExtensionState(enabled, value));
+
+    // Add or remove service extension from [enabledServiceExtensions].
+    if (enabled) {
+      _enabledServiceExtensions[name] =
+          new ServiceExtensionState(enabled, value);
+    } else {
+      _enabledServiceExtensions.remove(name);
+    }
+  }
+
+  StreamSubscription<bool> hasServiceExtension(String name, Function onData) {
+    final StreamController<bool> streamController =
+        _getServiceExtensionController(name);
+    return streamController.stream.listen(onData);
+  }
+
+  StreamSubscription<ServiceExtensionState> getServiceExtensionState(
+      String name, Function onData) {
+    final StreamController<ServiceExtensionState> streamController =
+        _getServiceExtensionStateController(name);
+    return streamController.stream.listen(onData);
+  }
+
+  StreamController<bool> _getServiceExtensionController(String name) {
+    return _getStream(name, _serviceExtensionController,
+        onFirstListenerSubscribed: () {
+      // If the service extension is in [_serviceExtensions], then we have been
+      // waiting for a listener to add the initial true event. Otherwise, the
+      // service extension is not available, so we should add a false event.
+      _serviceExtensionController[name].add(_serviceExtensions.contains(name));
+    });
+  }
+
+  StreamController<ServiceExtensionState> _getServiceExtensionStateController(
+      String name) {
+    return _getStream(name, _serviceExtensionStateController,
+        onFirstListenerSubscribed: () {
+      // If the service extension is enabled, add the current state as the first
+      // event. Otherwise, add a disabled state as the first event.
+      if (_enabledServiceExtensions.containsKey(name)) {
+        assert(_enabledServiceExtensions[name].enabled);
+        _serviceExtensionStateController[name]
+            .add(_enabledServiceExtensions[name]);
+      } else {
+        _serviceExtensionStateController[name]
+            .add(new ServiceExtensionState(false, null));
+      }
+    });
+  }
+
+  /// Initializes a generic stream if it does not already exist for the given
+  /// extension name.
+  StreamController<T> _getStream<T>(
+      String name, Map<String, StreamController<T>> streams,
+      {@required Function onFirstListenerSubscribed}) {
+    return streams[name] ??=
```

Comment:
I'd recommend using 
putIfAbsent for this case as the intent and behavior is a little clearer. Possible slight performance boost as well as one fewer map lookup has to be performed although the cost to allocate the closure probably outweighs that,.

`streams.putIfAbsent(name, () => StreamController<T>.broadcast(onListen: onFirstListenerSubscribed));`

Reply:
Done.